readme.md
# validate

Validate object properties in javascript.

[![npm version](http://img.shields.io/npm/v/validate.svg?style=flat)](https://npmjs.org/package/validate)
[![Build Status](http://img.shields.io/travis/eivindfjeldstad/validate.svg?style=flat)](https://travis-ci.org/eivindfjeldstad/validate)

## Usage

Define a schema and call `.validate()` with the object you want to validate.
The `.validate()` function returns an array of validation errors.

```js
const Schema = require('validate')

const user = new Schema({
  username: {
    type: 'string',
    required: true,
    length: { min: 3, max: 32 }
  },
  name: {
    type: 'string',
    required: true
  },
  pets: [{
    name: { type: 'string' },
    animal: { enum: ['cat', 'dog']}
  }],
  address: {
    street: {
      type: 'string',
      required: true
    },
    city: {
      type: 'string',
      required: true
    }
    zip: {
      type: 'string',
      match: /[0-9]+/,
      required: true
    }
  }
})

const errors = user.validate(obj)
```

Each error has a `.path`, describing the full path of the property that failed validation, and a `.message` describing the error.

```js
errors[0].path //=> 'address.street'
errors[0].message //=> 'address.street is required.'
```

### Custom error messages

You can override the default error messages by passing an object to `Schema#message()`.

```js
const post = new Schema({
  title: { required: true }
})

post.messsage({
  required: (path) => `${path} can not be empty.`
})

const [error] = post.validate({})
assert(error.message = 'title can not be empty.')
```

### Nesting

Objects and arrays can be nested as deep as you want:

```js
const event = new Schema({
  title: { type: 'string' },
  participants: [{
    name: {
      type: 'string'
    },
    email: {
      type: 'string',
      required: true
    },
    things: [{
      name: { type: 'string' },
      amount: { type: 'number' }
    }]
  }]
})
```

Arrays can be defined implicitly, like in the above example, or explicitly:

```js
const post = new Schema({
  title: { type: 'string' },
  content: { type: 'string' },
  keywords: {
    type: 'array',
    each: { type: 'string' }
  }
})
```

Nesting also works with schemas:

```js
const user = new Schema({
  name: { type: 'string' },
  email: { type: 'string' }
})

const post = new Schema({
  title: { type: 'string' },
  content: { type: 'string' },
  author: user
})
```

If you think it should work, it probably works.

### Custom validators

Custom validators can be defined by passing an object with named validators to `.use`:

```js
const hexColor = val => /^#[0-9a-fA-F]$/.test(val)

const car = new Schema({
  color: {
    type: 'string',
    use: { hexColor }
  }
})
```

Define a custom error message for the validator:

```js
car.message({
  hexColor: path => `${path} must be a valid color.`
})
```

### Chainable API

If you want to avoid constructing large objects, you can add paths to a schema by using the chainable API:

```js
const user = new Schema()

user
  .path('username')
    .type('string')
    .required()
  .path('address.zip')
    .type('string')
    .required()
```

Array elements can be defined by using `$` as a placeholder for indices:

```js
const user = new Schema()

user
  .path('pets')
    .type('array')
  .path('pets.$')
    .type('string')
```

This is equivalent to writing

```js
const user = new Schema({ pets: [{ type: 'string' }]})
```

### Typecasting

Values can be automatically typecast before validation.
To enable typecasting, pass an options object to the `Schema` constructor with `typecast` set to `true`.

```js
const user = new Schema(definition, { typecast: true })
```

You can override this setting by passing an option to `.validate()`.

```js
user.validate(obj, { typecast: false })
```

### Property stripping

By default, all values not defined in the schema will be stripped from the object.
Set `.strip = false` on the options object to disable this behavior.

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [Property](#property)
    -   [schema](#schema)
    -   [use](#use)
    -   [required](#required)
    -   [type](#type)
    -   [length](#length)
    -   [enum](#enum)
    -   [match](#match)
    -   [each](#each)
    -   [path](#path)
    -   [typecast](#typecast)
    -   [validate](#validate)
-   [Schema](#schema-1)
    -   [path](#path-1)
    -   [validate](#validate-1)
    -   [assert](#assert)
    -   [message](#message)
    -   [validator](#validator)

### Property

A property instance gets returned whenever you call `schema.path()`.
Properties are also created internally when an object is passed to the Schema constructor.

**Parameters**

-   `name` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** the name of the property
-   `schema` **[Schema](#schema)** parent schema

#### schema

Mount given `schema` on current path.

**Parameters**

-   `schema` **[Schema](#schema)** the schema to mount

**Examples**

```javascript
const user = new Schema({ email: 'string' })
prop.schema(user)
```

Returns **[Property](#property)**

#### use

Validate using named functions from the given object.
Error messages can be defined by providing an object with
named error messages/generators to `schema.message()`

The message generator receives the value being validated,
the object it belongs to and any additional arguments.

**Parameters**

-   `fns` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** object with named validation functions to call

**Examples**

```javascript
const schema = new Schema()
const prop = schema.path('some.path')

schema.message({
  binary: (path, ctx) => `${path} must be binary.`,
  bits: (path, ctx, bits) => `${path} must be ${bits}-bit`
})

prop.use({
  binary: (val, ctx) => /^[01]+$/i.test(val),
  bits: [(val, ctx, bits) => val.length == bits, 32]
})
```

Returns **[Property](#property)**

#### required

Registers a validator that checks for presence.

**Parameters**

-   `bool` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** `true` if required, `false` otherwise (optional, default `true`)

**Examples**

```javascript
prop.required()
```

Returns **[Property](#property)**

#### type

Registers a validator that checks if a value is of a given `type`

**Parameters**

-   `type` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** type to check for

**Examples**

```javascript
prop.type('string')
```

Returns **[Property](#property)**

#### length

Registers a validator that checks length.

**Parameters**

-   `rules` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** object with `.min` and `.max` properties
    -   `rules.min` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** minimum length
    -   `rules.max` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** maximum length

**Examples**

```javascript
prop.length({ min: 8, max: 255 })
```

Returns **[Property](#property)**

#### enum

Registers a validator for enums.

**Parameters**

-   `enums`  
-   `rules` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)** allowed values

**Examples**

```javascript
prop.enum(['cat', 'dog'])
```

Returns **[Property](#property)**

#### match

Registers a validator that checks if a value matches given `regexp`.

**Parameters**

-   `regexp` **[RegExp](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/RegExp)** regular expression to match

**Examples**

```javascript
prop.match(/some\sregular\sexpression/)
```

Returns **[Property](#property)**

#### each

Registers a validator that checks each value in an array against given `rules`.

**Parameters**

-   `rules` **([Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array) \| [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object) \| [Schema](#schema) \| [Property](#property))** rules to use

**Examples**

```javascript
prop.each({ type: 'string' })
prop.each([{ type: 'number' }])
prop.each({ things: [{ type: 'string' }]})
prop.each(schema)
```

Returns **[Property](#property)**

#### path

Proxy method for schema path. Makes chaining properties together easier.

**Parameters**

-   `args` **...any**

**Examples**

```javascript
schema
  .path('name')
    .type('string')
    .required()
  .path('email')
    .type('string')
    .required()
```

#### typecast

Typecast given `value`

**Parameters**

-   `value` **Mixed** value to typecast

**Examples**

```javascript
prop.type('string')
prop.typecast(123) // => '123'
```

Returns **Mixed**

#### validate

Validate given `value`

**Parameters**

-   `value` **Mixed** value to validate
-   `ctx` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** the object containing the value
-   `path` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** path of the value being validated (optional, default `this.name`)

**Examples**

```javascript
prop.type('number')
assert(prop.validate(2) == false)
assert(prop.validate('hello world') instanceof Error)
```

Returns **([Error](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Error) \| [Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean))**

### Schema

A Schema defines the structure that objects should be validated against.

**Parameters**

-   `obj` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)?** schema definition (optional, default `{}`)
-   `opts` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)?** options (optional, default `{}`)
    -   `opts.typecast` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** typecast values before validation (optional, default `false`)
    -   `opts.strip` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** strip properties not defined in the schema (optional, default `true`)

**Examples**

```javascript
const post = new Schema({
  title: {
    type: 'string',
    required: true,
    length: { min: 1, max: 255 }
  },
  content: {
    type: 'string',
    required: true
  },
  published: {
    type: 'date',
    required: true
  },
  keywords: [{ type: 'string' }]
})
```

```javascript
const author = new Schema({
  name: {
    type: 'string',
    required: true
  },
  email: {
    type: 'string',
    required: true
  },
  posts: [post]
})
```

#### path

Create or update `path` with given `rules`.

**Parameters**

-   `path` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** full path using dot-notation
-   `rules` **([Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object) \| [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array) \| [String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Schema](#schema) \| [Property](#property))?** rules to apply

**Examples**

```javascript
const schema = new Schema()
schema.path('name.first', { type: 'string' })
schema.path('name.last').type('string').required()
```

Returns **[Property](#property)**

#### validate

Validate given `obj`.

**Parameters**

-   `obj` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** the object to validate
-   `opts` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)?** options, see [Schema](#schema-1) (optional, default `{}`)

**Examples**

```javascript
const schema = new Schema({ name: { required: true }})
const errors = schema.validate({})
assert(errors.length == 1)
assert(errors[0].message == 'name is required')
assert(errors[0].path == 'name')
```

Returns **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)**

#### assert

Assert that given `obj` is valid.

**Parameters**

-   `obj` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)**
-   `opts` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)?**

**Examples**

```javascript
const schema = new Schema({ name: 'string' })
schema.assert({ name: 1 }) // => Throws an error
```

#### message

Override default error messages.

**Parameters**

-   `name` **([String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object))** name of the validator or an object with name-message pairs
-   `message` **([String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function))?** the message or message generator to use

**Examples**

```javascript
const hex = (val) => /^0x[0-9a-f]+$/.test(val)
schema.path('some.path').use({ hex })
schema.message('hex', path => `${path} must be hexadecimal`)
```

```javascript
schema.message({ hex: path => `${path} must be hexadecimal` })
```

Returns **[Schema](#schema)**

#### validator

Override default validators.

**Parameters**

-   `name` **([String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object))** name of the validator or an object with name-function pairs
-   `fn` **[Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)?** the function to use

**Examples**

```javascript
schema.validator('required', val => val != null)
```

```javascript
schema.validator({ required: val => val != null })
```

Returns **[Schema](#schema)**

## Licence

MIT


History.md

3.1.0 / 2018-03-07
==================

  * version 3.1.0
  * update readme
  * add length validator
  * remove component.json and Makefile
  * fix typo
  * update readme
  * Allow type shorthand, closes #18

3.0.1 / 2015-09-24
==================

  * require should respect boolean argument
  * typo
  * Update component.json
  * add History.md

3.0.0 / 2015-09-24
==================

  * remove component/assert from dependencies
  * add license to package.json
  * return error objects instead of strings
  * add new node versions to travis.md
  * fix build status in readme.md
  * add `property.each()`

2.1.6 / 2014-06-02
==================

  * fix stripping of nested objects. closes #15
  * update readme

2.1.5 / 2014-05-25
==================

  * fix typecasting of undefineds

2.1.4 / 2014-05-25
==================

  * refactor

2.1.3 / 2014-05-12
==================

  * update readme
  * use semver dependencies. closes #14
  * use svg badge
  * fix deps
  * refactor
  * add assert as dep
  * fix component.json
  * add note about typecasting

2.1.2 / 2014-01-14
==================

  * use typecast

2.1.1 / 2013-12-23
==================

  * fix typecast test
  * refactor


lib/property.js
const typeOf = require('component-type');
const typecast = require('typecast');

/**
 * A property instance gets returned whenever you call `schema.path()`.
 * Properties are also created internally when an object is passed to the Schema constructor.
 *
 * @param {String} name - the name of the property
 * @param {Schema} schema - parent schema
 */

class Property {
  constructor(name, schema) {
    this.name = name;
    this.registry = {};
    this._schema = schema;
    this._type = null;
    this.messages = schema.messages;
    this.validators = schema.validators;
  }

  /**
   * Mount given `schema` on current path.
   *
   * @example
   * const user = new Schema({ email: 'string' })
   * prop.schema(user)
   *
   * @param {Schema} schema - the schema to mount
   * @return {Property}
   */

  schema(schema) {
    this._schema.path(this.name, schema);
    return this;
  }

  /**
   * Validate using named functions from the given object.
   * Error messages can be defined by providing an object with
   * named error messages/generators to `schema.message()`
   *
   * The message generator receives the value being validated,
   * the object it belongs to and any additional arguments.
   *
   * @example
   * const schema = new Schema()
   * const prop = schema.path('some.path')
   *
   * schema.message({
   *   binary: (path, ctx) => `${path} must be binary.`,
   *   bits: (path, ctx, bits) => `${path} must be ${bits}-bit`
   * })
   *
   * prop.use({
   *   binary: (val, ctx) => /^[01]+$/i.test(val),
   *   bits: [(val, ctx, bits) => val.length == bits, 32]
   * })
   *
   * @param {Object} fns - object with named validation functions to call
   * @return {Property}
   */

  use(fns) {
    Object.keys(fns).forEach(name => {
      let arr = fns[name];
      if (!Array.isArray(arr)) arr = [arr];
      const fn = arr.shift();
      this._register(name, arr, fn);
    })

    return this;
  }

  /**
   * Registers a validator that checks for presence.
   *
   * @example
   * prop.required()
   *
   * @param {Boolean} [bool] - `true` if required, `false` otherwise
   * @return {Property}
   */

  required(bool = true) {
    return this._register('required', [bool]);
  }

  /**
   * Registers a validator that checks if a value is of a given `type`
   *
   * @example
   * prop.type('string')
   *
   * @param {String} type - type to check for
   * @return {Property}
   */

  type(type) {
    this._type = type;
    return this._register('type', [type]);
  }

  /**
   * Registers a validator that checks length.
   *
   * @example
   * prop.length({ min: 8, max: 255 })
   *
   * @param {Object} rules - object with `.min` and `.max` properties
   * @param {Number} rules.min - minimum length
   * @param {Number} rules.max - maximum length
   * @return {Property}
   */

  length(rules) {
    return this._register('length', [rules]);
  }

  /**
   * Registers a validator for enums.
   *
   * @example
   * prop.enum(['cat', 'dog'])
   *
   * @param {Array} rules - allowed values
   * @return {Property}
   */

  enum(enums) {
    return this._register('enum', [enums]);
  }

  /**
   * Registers a validator that checks if a value matches given `regexp`.
   *
   * @example
   * prop.match(/some\sregular\sexpression/)
   *
   * @param {RegExp} regexp - regular expression to match
   * @return {Property}
   */

  match(regexp) {
    return this._register('match', [regexp]);
  }

  /**
   * Registers a validator that checks each value in an array against given `rules`.
   *
   * @example
   * prop.each({ type: 'string' })
   * prop.each([{ type: 'number' }])
   * prop.each({ things: [{ type: 'string' }]})
   * prop.each(schema)
   *
   * @param {Array|Object|Schema|Property} rules - rules to use
   * @return {Property}
   */

  each(rules) {
    this._schema.path(`${this.name}.$`, rules);
    return this;
  }

  /**
   * Proxy method for schema path. Makes chaining properties together easier.
   *
   * @example
   * schema
   *   .path('name')
   *     .type('string')
   *     .required()
   *   .path('email')
   *     .type('string')
   *     .required()
   *
   */

  path(...args) {
    return this._schema.path(...args);
  }

  /**
   * Typecast given `value`
   *
   * @example
   * prop.type('string')
   * prop.typecast(123) // => '123'
   *
   * @param {Mixed} value - value to typecast
   * @return {Mixed}
   */

  typecast(value) {
    if (!this._type) return value;
    return typecast(value, this._type);
  }

  /**
   * Validate given `value`
   *
   * @example
   * prop.type('number')
   * assert(prop.validate(2) == false)
   * assert(prop.validate('hello world') instanceof Error)
   *
   * @param {Mixed} value - value to validate
   * @param {Object} ctx - the object containing the value
   * @param {String} [path] - path of the value being validated
   * @return {Error|Boolean}
   */

  validate(value, ctx, path = this.name) {
    const types = Object.keys(this.registry);
    const done = {};
    let error;

    // Required first
    if (error = this._run('required', value, ctx, path)) {
      return error;
    }

    // No need to continue if value is null-ish
    if (value == null) return false;

    // Run type second
    if (error = this._run('type', value, ctx, path)) {
      return error;
    }

    // Make sure required and run are not executed again
    done.required = true;
    done.type = true;

    // Run the rest
    for (let type of types) {
      if (done[type]) continue;
      if (error = this._run(type, value, ctx, path)) {
        return error;
      }
    }

    return false;
  }

  /**
   * Run validator of given `type`
   *
   * @param {String} type - type of validator
   * @param {Mixed} value - value to validate
   * @param {Object} ctx - the object containing the value
   * @param {String} path - path of the value being validated
   * @return {Error}
   * @private
   */

  _run(type, value, ctx, path) {
    if (!this.registry[type]) return;
    const {args, fn} = this.registry[type];
    let validator = fn || this.validators[type];
    let valid = validator(value, ctx, ...args);
    if (!valid) return this._error(type, ctx, args, path);
  }

  /**
   * Register validator
   *
   * @param {String} type - type of validator
   * @param {Array} args - argument to pass to validator
   * @param {Function} [fn] - custom validation function to call
   * @return {Property}
   * @private
   */

  _register(type, args, fn) {
    this.registry[type] = { args, fn }
    return this;
  }

  /**
   * Create an error
   *
   * @param {String} type - type of validator
   * @param {Object} ctx - the object containing the value
   * @param {Array} args - arguments to pass
   * @param {String} path - path of the value being validated
   * @return {Error}
   * @private
   */

  _error(type, ctx, args, path) {
    let message = this.messages[type] || this.messages.default;

    if (typeof message == 'function') {
      message = message(path, ctx, ...args);
    }

    const err = new Error(message);
    err.path = path;
    return err;
  }
}

module.exports = Property;
